### PR TITLE
Fix S2699 throws AD0001 exception

### LINKED ISSUE #4332: S2699 throws AD0001 when invoked via AnalyzerRunner.exe
happens (at least) for `nodatime`, `n2cms`, `efcore`

```
  AD0001: 16 instances
error AD0001: Analyzer 'SonarAnalyzer.Rules.CSharp.TestMethodShouldContainAssertion' threw an exception of type 'System.ArgumentException' with message 'SyntaxTree is not part of the compilation
Parameter name: syntaxTree'.
Exception occurred with following context:
Compilation: NodaTime.Test
SyntaxTree: C:\Workspace\external-projects\PERF_SUITE\projects\nodatime\src\NodaTime.Test\Globalization\NodaFormatInfoTest.cs
SyntaxNode: [Test] [TestCaseSource(typeof(Cultures ... [MethodDeclarationSyntax]@[756..967) (21,8)-(26,9)

System.ArgumentException: SyntaxTree is not part of the compilation
Parameter name: syntaxTree
   at Microsoft.CodeAnalysis.CSharp.CSharpCompilation.GetSemanticModel(SyntaxTree syntaxTree, Boolean ignoreAccessibility) in C:\Workspace\external-projects\roslyn\src\Compilers\CSharp\Portable\Compilation\CSharpCompilation.cs:line 2147
   at Microsoft.CodeAnalysis.CSharp.CSharpCompilation.CommonGetSemanticModel(SyntaxTree syntaxTree, Boolean ignoreAccessibility) in C:\Workspace\external-projects\roslyn\src\Compilers\CSharp\Portable\Compilation\CSharpCompilation.cs:line 3229
   at Microsoft.CodeAnalysis.Compilation.GetSemanticModel(SyntaxTree syntaxTree, Boolean ignoreAccessibility) in C:\Workspace\external-projects\roslyn\src\Compilers\Core\Portable\Compilation\Compilation.cs:line 211
   at SonarAnalyzer.Rules.CSharp.TestMethodShouldContainAssertion.ContainsAssertion(MethodDeclarationSyntax methodDeclaration, SemanticModel previousSemanticModel, ISet`1 visitedSymbols, Int32 level)
   at SonarAnalyzer.Rules.CSharp.TestMethodShouldContainAssertion.ContainsAssertion(MethodDeclarationSyntax methodDeclaration, SemanticModel previousSemanticModel, ISet`1 visitedSymbols, Int32 level)
   at SonarAnalyzer.Rules.CSharp.TestMethodShouldContainAssertion.<>c.<Initialize>b__11_0(SyntaxNodeAnalysisContext c)
   at SonarAnalyzer.Helpers.DiagnosticAnalyzerContextHelper.<>c__DisplayClass1_0`1.<RegisterSyntaxNodeActionInNonGenerated>b__0(SyntaxNodeAnalysisContext c)
   at SonarAnalyzer.Helpers.SonarAnalysisContext.<>c__DisplayClass46_0`1.<RegisterContextAction>b__0(TContext c)
   at Microsoft.CodeAnalysis.Diagnostics.AnalyzerExecutor.<>c__62`1.<ExecuteSyntaxNodeAction>b__62_0(ValueTuple`2 data) in C:\Workspace\external-projects\roslyn\src\Compilers\Core\Portable\DiagnosticAnalyzer\AnalyzerExecutor.cs:line 918
   at Microsoft.CodeAnalysis.Diagnostics.AnalyzerExecutor.ExecuteAndCatchIfThrows_NoLock[TArg](DiagnosticAnalyzer analyzer, Action`1 analyze, TArg argument, Nullable`1 info) in C:\Workspace\external-projects\roslyn\src\Compilers\Core\Portable\DiagnosticAnalyzer\AnalyzerExecutor.cs:line 1596
-----

Suppress the following diagnostics to disable this analyzer: S2699
```

### PR DESCRIPTION
**Pull request by dmitry-golovinov-sonarsource**
Fixes #4332

### CODE REVIEW

File: analyzers/src/SonarAnalyzer.Common/Trackers/ConstantValueFinder.cs [hunk 1]
```diff
@@ -46,10 +46,19 @@ protected ConstantValueFinder(SemanticModel semanticModel, AssignmentFinder assi
...
+            if (node == null || node.RawKind == nullLiteralExpressionSyntaxKind) // Performance shortcut
+            {
+                return null;
+            }
+            var nodeSemanticModel = node.EnsureCorrectSemanticModelOrDefault(semanticModel);
+            if (nodeSemanticModel == null)
+            {
+                return null;
+            }
+            return nodeSemanticModel.GetConstantValue(node).Value ?? FindAssignedConstant(node, visitedVariables);
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/Trackers/ConstantValueFinder.cs:60]:
```suggestion
            return node.EnsureCorrectSemanticModelOrDefault(semanticModel) is { } nodeSemanticModel
                ? nodeSemanticModel.GetConstantValue(node).Value ?? FindAssignedConstant(node, visitedVariables)
                : null;
```

File: analyzers/src/SonarAnalyzer.CSharp/Rules/MethodParameterUnused.cs [hunk 2]
```diff
@@ -236,7 +236,7 @@ private static IImmutableList<IParameterSymbol> GetUnusedParameters(MethodContex
         private static bool IsMethodUsedAsEventHandlerFunctionOrActionInExpression(IMethodSymbol methodSymbol, ExpressionSyntax expression, SemanticModel semanticModel) =>
             !expression.IsKind(SyntaxKind.InvocationExpression)
             && IsStandaloneExpression(expression)
-            && methodSymbol.Equals(semanticModel.GetSymbolInfo(expression).Symbol?.OriginalDefinition);
+            && methodSymbol.Equals(semanticModel?.GetSymbolInfo(expression).Symbol?.OriginalDefinition);
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/MethodParameterUnused.cs:239]: Let's add `&& semanticModel != null` as 2nd condition for performance reason. We don't need to evaluate `IsStandaloneExpression` in that case

**Review by pavel-mikula-sonarsource — CHANGES_REQUESTED** [review]
Improvements suggested

**Review by pavel-mikula-sonarsource — APPROVED** [review]
LGTM

File: analyzers/src/SonarAnalyzer.CSharp/Rules/TestMethodShouldContainAssertion.cs [hunk 3]
```diff
@@ -93,7 +93,11 @@ public sealed class TestMethodShouldContainAssertion : SonarDiagnosticAnalyzer
 
         private static bool ContainsAssertion(MethodDeclarationSyntax methodDeclaration, SemanticModel previousSemanticModel, ISet<IMethodSymbol> visitedSymbols, int level)
         {
-            var currentSemanticModel = methodDeclaration.EnsureCorrectSemanticModel(previousSemanticModel);
+            var currentSemanticModel = methodDeclaration.EnsureCorrectSemanticModelOrDefault(previousSemanticModel);
+            if (currentSemanticModel == null)
+            {
+                return false;
```
Comment (andrei-epure-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/TestMethodShouldContainAssertion.cs:99]: won't this mean we're going to raise issues when there is this problem?  If `!ContainsAssertion` , we raise.

We should be silent if there's no semantic model. Or am I missing something?

**Review by andrei-epure-sonarsource — DISMISSED** [review]

**Comment by andrei-epure-sonarsource** [comment]
Let's not merge this before SQ LTS. And try to come up with a tests to cover the logic.

The `analyzerRunner.exe` scenario is an edge case, no need to hurry.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/TestMethodShouldContainAssertion.cs (hunk 3 above)
Comment (andrei-epure-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/TestMethodShouldContainAssertion.cs:99]: from offline discussion:
>It's in crossmethod inspection, we return false when we dont't inspect deeper (or we can't tell). As it's not for real life scenario, it should be fine as is

File: analyzers/src/SonarAnalyzer.CSharp/Helpers/CSharpSymbolUsageCollector.cs [hunk 4]
```diff
@@ -64,8 +64,11 @@ public CSharpSymbolUsageCollector(Compilation compilation, IEnumerable<ISymbol>
 
         public override void Visit(SyntaxNode node)
         {
-            semanticModel = node.EnsureCorrectSemanticModel(semanticModel ?? compilation.GetSemanticModel(node.SyntaxTree));
-            base.Visit(node);
+            semanticModel = node.EnsureCorrectSemanticModelOrDefault(semanticModel ?? compilation.GetSemanticModel(node.SyntaxTree));
+            if (semanticModel != null)
```
Comment (andrei-epure-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Helpers/CSharpSymbolUsageCollector.cs:68]: For the C# usage symbol collector, you can create two different "snippet" compilations in the UT, and give compilation1 to the constructor, and give a SyntaxNode from the second compilation to Visit. Then, it should not do the visiting.

See ConstantValueFinderTest for example how to do that

And then check `UsedSymbols` or some other property, that it collects symbols when the semanticModel is resolved, and that it doesn't when the SemanticModel doesn't get resolved.

**Review by pavel-mikula-sonarsource — CHANGES_REQUESTED** [review]
Let's add UT as discussed. I'm adding `Request changes` to keep the history explicit.

File: analyzers/tests/SonarAnalyzer.UnitTest/Trackers/ConstantValueFinderTest.cs [hunk 5]
```diff
@@ -57,5 +57,34 @@ public int Method()
...
+        return Field;
+    }
+}";
+            var compilation1 = SolutionBuilder.Create().AddProject(AnalyzerLanguage.CSharp, createExtraEmptyFile: false)
+                .AddSnippet(code1)
+                .GetCompilation();
+            var compilation2 = SolutionBuilder.Create().AddProject(AnalyzerLanguage.CSharp, createExtraEmptyFile: false)
+                .AddSnippet(code2)
+                .GetCompilation();
+            var tree = compilation2.SyntaxTrees.Single(x => x.GetRoot().DescendantNodes().OfType<MethodDeclarationSyntax>().Any());
```
Comment (andrei-epure-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/Trackers/ConstantValueFinderTest.cs:84]: I suggest to improve the naming a bit, to make the differences obvious and easier to follow:
- compilation1  -> firstCompilation
- compilation2 -> secondCompilation`

You don't need `tree` from what I see, you can merge that into the LINQ query for returnExpression
Also:
-  returnExpression -> secondCompilationReturnExpression

And then you have:
```
firstCompilationFinder.FindConstant(secondCompilationReturnExpression).Should().BeNull();
```

which is much clearer

File: analyzers/tests/SonarAnalyzer.UnitTest/Trackers/ConstantValueFinderTest.cs [hunk 6]
```diff
@@ -57,5 +57,34 @@ public int Method()
             var finder = new CSharpConstantValueFinder(compilation.GetSemanticModel(tree));
             finder.FindConstant(returnExpression).Should().Be(42);
         }
+
+        [TestMethod]
+        public void WrongCompilationBeingUsed()
+        {
+            const string code1 = @"
+public partial class Sample
```
Comment (andrei-epure-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/Trackers/ConstantValueFinderTest.cs:65]: using `partial class` and the same name makes the test hard to follow, because the two partial classes are in two different solutions, so they're totally different.

IMO for the sake of testing this scenario you can have just two classes `Foo` and `Bar`  and pass the wrong Compilation to the class, and it will be easy to understand what this test is doing

File: analyzers/tests/SonarAnalyzer.UnitTest/Helpers/CSharpSymbolUsageCollectorTest.cs [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+}";
+            const string code2 = @"
+public partial class Sample
+{
+    public int Bar()
+    {
+        return Field;
+    }
+}";
+            var compilation1 = SolutionBuilder.Create().AddProject(AnalyzerLanguage.CSharp, createExtraEmptyFile: false)
```
Comment (andrei-epure-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/Helpers/CSharpSymbolUsageCollectorTest.cs:58]: same here: I suggest to use "firstCompilation" and "secondCompilation" (and "firstTree" and "secondTree") because it's easier do differentiate the names

File: analyzers/tests/SonarAnalyzer.UnitTest/Helpers/CSharpSymbolUsageCollectorTest.cs [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+namespace SonarAnalyzer.UnitTest.Helpers
+{
+    [TestClass]
+    public class CSharpSymbolUsageCollectorTest
+    {
+        [TestMethod]
+        public void VerifyUsagesBeingCollectedOnMatchingSyntaxNodes()
+        {
+            const string code1 = @"
+public partial class Sample
```
Comment (andrei-epure-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/Helpers/CSharpSymbolUsageCollectorTest.cs:40]: same comment as on `WrongCompilationBeingUsed` - why use `partial class Sample` in both ?

either way, in real life compilation2  wouldn't compile in its own solution, right?

**Review by andrei-epure-sonarsource — CHANGES_REQUESTED** [review]
Great to see these tests, I have some readability suggestions

File: analyzers/tests/SonarAnalyzer.UnitTest/Helpers/CSharpSymbolUsageCollectorTest.cs (hunk 8 above)
Comment (dmitry-golovinov-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/Helpers/CSharpSymbolUsageCollectorTest.cs:40]: Renamed the classes and removed partial

File: analyzers/tests/SonarAnalyzer.UnitTest/Trackers/ConstantValueFinderTest.cs [hunk 9]
```diff
@@ -57,5 +57,33 @@ public int Method()
...
+            const string firstSnippet = @"
+public class Foo
+{
+    private static int Original = 42;
+    private int Field = Original;
+}";
+            const string secondSnippet = @"
+public class Bar
+{
+    public int Method()
```
Comment (andrei-epure-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/Trackers/ConstantValueFinderTest.cs:73]: this code still doesn't compile

So even if you use a secondCompilationFinder, it will still return `null`, because the `Field` symbol does not have any symbol.

If I add these to the test, it fails, because there's no symbol:
```
var secondCompilationFinder = new CSharpConstantValueFinder(secondCompilation.GetSemanticModel(secondCompilation.SyntaxTrees.Single()));
secondCompilationFinder.FindConstant(secondCompilationReturnExpression).Should().NotBeNull();
```

This brings confusion to the test (is it just testing that the `firstCompilationFinder` cannot find symbols because the wrong compilation is used, or because the `secondSnippet` doesn't compile , so it has no symbols even for its own compilation?).

I suggest to make the `Bar` class compilable.

```
public class Bar
{
    private int Field = 42;
    public int Method()
    {
      return Field;
     }
}
```
And then these assertions will both work:
```
firstCompilationFinder.FindConstant(secondCompilationReturnExpression).Should().BeNull();
secondCompilationFinder.FindConstant(secondCompilationReturnExpression).Should().NotBeNull();
```

File: analyzers/tests/SonarAnalyzer.UnitTest/Helpers/CSharpSymbolUsageCollectorTest.cs [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+    public int FooMethod(int arg)
+    {
+        Field += arg;
+        return Field;
+    }
+}";
+            const string secondSnippet = @"
+public class Bar
+{
+    public int BarMethod()
```
Comment (andrei-epure-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/Helpers/CSharpSymbolUsageCollectorTest.cs:53]: same here, please add `private int Field = 42;` above `BarMethod` so that the code can compile

File: analyzers/tests/SonarAnalyzer.UnitTest/Helpers/CSharpSymbolUsageCollectorTest.cs [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+            var secondCompilation = SolutionBuilder.Create().AddProject(AnalyzerLanguage.CSharp, createExtraEmptyFile: false)
+                .AddSnippet(secondSnippet)
+                .GetCompilation();
+
+            var firstTree = firstCompilation.SyntaxTrees.Single();
+            var fooMethodDecl = firstTree.GetRoot().DescendantNodes().OfType<MethodDeclarationSyntax>().Single();
+            var semanticModel = firstCompilation.GetSemanticModel(firstTree);
+
+            var fieldSymbol = semanticModel.GetSymbolInfo(fooMethodDecl.DescendantNodes().OfType<ReturnStatementSyntax>().Single().Expression).Symbol;
+            var knownSymbolsFirstCompilation = new List<ISymbol> { fieldSymbol };
```
Comment (andrei-epure-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/Helpers/CSharpSymbolUsageCollectorTest.cs:70]: nitpick: if you use `firstCompilationKnownSymbols` I can tell from the beginning that this refers to `first...` - the same for other names. It's easier to navigate the test, I can see `var firstCompilationXyz` and I don't have to read what's on the rest of the line to tell what the `Xyz` is about.

**Review by andrei-epure-sonarsource — CHANGES_REQUESTED** [review]
last few comments

File: analyzers/tests/SonarAnalyzer.UnitTest/Trackers/ConstantValueFinderTest.cs (hunk 9 above)
Comment (dmitry-golovinov-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/Trackers/ConstantValueFinderTest.cs:73]: Haven't got this from the beginning. Indeed, those 2 are not compilable!

**Review by andrei-epure-sonarsource — APPROVED** [review]
LGTM

File: analyzers/tests/SonarAnalyzer.UnitTest/Helpers/CSharpSymbolUsageCollectorTest.cs [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+}";
+            var firstCompilation = SolutionBuilder.Create().AddProject(AnalyzerLanguage.CSharp, createExtraEmptyFile: false)
+                .AddSnippet(firstSnippet)
+                .GetCompilation();
+            var secondCompilation = SolutionBuilder.Create().AddProject(AnalyzerLanguage.CSharp, createExtraEmptyFile: false)
+                .AddSnippet(secondSnippet)
+                .GetCompilation();
+
+            var firstTree = firstCompilation.SyntaxTrees.Single();
+            var fooMethodDecl = firstTree.GetRoot().DescendantNodes().OfType<MethodDeclarationSyntax>().Single();
```
Comment (pavel-mikula-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/Helpers/CSharpSymbolUsageCollectorTest.cs:66]: _Nitpick:_ This should be enough to name it, as there's no other `fooMethodSymbol` or `fooMethodBody` that would collide. We usually don't use half-words like the Java `Impl`
```suggestion
            var fooMethod = firstTree.GetRoot().DescendantNodes().OfType<MethodDeclarationSyntax>().Single();
```

File: analyzers/tests/SonarAnalyzer.UnitTest/Helpers/CSharpSymbolUsageCollectorTest.cs [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+            var firstCompilationSemanticModel = firstCompilation.GetSemanticModel(firstTree);
+
+            var firstCompilationFieldSymbol = firstCompilationSemanticModel.GetSymbolInfo(fooMethodDecl.DescendantNodes().OfType<ReturnStatementSyntax>().Single().Expression).Symbol;
```
Comment (pavel-mikula-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/Helpers/CSharpSymbolUsageCollectorTest.cs:69]: _Nitpick:_ I don't see a reason for this empty line. It's a sequence of `firstCompilationXxxx =` initializations.
```suggestion
            var firstCompilationSemanticModel = firstCompilation.GetSemanticModel(firstTree);
            var firstCompilationFieldSymbol = firstCompilationSemanticModel.GetSymbolInfo(fooMethodDecl.DescendantNodes().OfType<ReturnStatementSyntax>().Single().Expression).Symbol;
```

File: analyzers/tests/SonarAnalyzer.UnitTest/Helpers/CSharpSymbolUsageCollectorTest.cs [hunk 14]
```diff
@@ -0,0 +1,85 @@
...
+            var firstCompilationKnownSymbols = new List<ISymbol> { firstCompilationFieldSymbol };
+
+            // compilation matches semantic model and syntax node
+            var firstCompilationUsageCollector = new CSharpSymbolUsageCollector(firstCompilation, firstCompilationKnownSymbols);
+            firstCompilationUsageCollector.Visit(fooMethodDecl);
+            firstCompilationUsageCollector.UsedSymbols.Should().NotBeEmpty();
+            var firstCompilationUsedSymbols = firstCompilationUsageCollector.UsedSymbols;
+
+            var secondTree = secondCompilation.SyntaxTrees.Single();
+            var barMethodDecl = secondTree.GetRoot().DescendantNodes().OfType<MethodDeclarationSyntax>().Single();
```
Comment (pavel-mikula-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/Helpers/CSharpSymbolUsageCollectorTest.cs:79]: Same as above
```suggestion
            var barMethod = secondTree.GetRoot().DescendantNodes().OfType<MethodDeclarationSyntax>().Single();
```

File: analyzers/tests/SonarAnalyzer.UnitTest/Helpers/CSharpSymbolUsageCollectorTest.cs [hunk 15]
```diff
@@ -0,0 +1,85 @@
...
+            // compilation doesn't match syntax node, since it belongs to another compilation
```
Comment (pavel-mikula-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/Helpers/CSharpSymbolUsageCollectorTest.cs:80]: _Nitpick:_ I'd move this comment 2 lines higher to have a regular structure (first comment, var, visit, assert, second comment, var, visit, assert).

**Review by pavel-mikula-sonarsource — APPROVED** [review]
LGTM. I left some nitpicks
